### PR TITLE
feat: recall and F1-score metrics for classification

### LINKED ISSUE #186: F1 score metric for classification
### Is your feature request related to a problem?

.

### Desired solution

.

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### LINKED ISSUE #187: Recall metric for classification
### Is your feature request related to a problem?

Computing the recall of a classifier for a specific class is a common task.

### Desired solution

* New method called `recall` in `Classifier`
* Required parameter `positive_class` to specify which value should be treated as positive (all other classes are treated as negative)
* Call [`sklearn.metrics.recall`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html#sklearn.metrics.recall_score)

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by PhilipGutberlet**
Closes #187.
Closes #186 .

### Summary of Changes

Added recall and F1-score functions to the _classifier.

Co-authered-by: 128832338+patrikguempel@users.noreply.github.com

### CODE REVIEW

**Comment by lars-reimann** [comment]
## [🦙 MegaLinter](https://megalinter.io/6.22.2) status: ❌ [ERROR](https://github.com/Safe-DS/Stdlib/actions/runs/4893704800)

| Descriptor  |                                 Linter                                 |Files|Fixed|                            Errors                            |Elapsed time|
|-------------|------------------------------------------------------------------------|-----|----:|--------------------------------------------------------------|------------|
|✅ PYTHON    |[black](https://megalinter.io/6.22.2/descriptors/python_black)          |    2|    1|                                                             0|0.95s       |
|✅ PYTHON    |[mypy](https://megalinter.io/6.22.2/descriptors/python_mypy)            |    2|     |                                                             0|1.69s       |
|❌ PYTHON    |[ruff](https://megalinter.io/6.22.2/descriptors/python_ruff)            |    2|    1|[1](https://github.com/Safe-DS/Stdlib/actions/runs/4893704800)|0.04s       |
|✅ REPOSITORY|[git_diff](https://megalinter.io/6.22.2/descriptors/repository_git_diff)|yes  |     |no                                                            |0.02s       |

See detailed report in [MegaLinter reports](https://github.com/Safe-DS/Stdlib/actions/runs/4893704800)
_Set `VALIDATE_ALL_CODEBASE: true` in mega-linter.yml to validate all sources, not only the diff_

_MegaLinter is graciously provided by [OX Security](https://www.ox.security/?ref=megalinter)_

File: tests/safeds/ml/classical/classification/test_classifier.py [hunk 1]
```diff
@@ -284,3 +284,104 @@ def test_should_return_1_if_never_expected_to_be_positive(self) -> None:
...
+        table = Table.from_dict(
+            {
+                "predicted": [1, 1, 0, 2],
+                "expected": [1, 0, 1, 2],
+            },
+        ).tag_columns(target_name="expected")
+
+        f1 = DummyClassifier().f1_score(table, 1)
+        assert f1 <= 0.667
+        assert f1 > 0.666
```
Comment (lars-reimann) [inline tests/safeds/ml/classical/classification/test_classifier.py]: Have a look at the [pytest.approx](https://docs.pytest.org/en/latest/reference/reference.html#pytest.approx) function for the comparison of floats.

File: tests/safeds/ml/classical/classification/test_classifier.py [hunk 2]
```diff
@@ -284,3 +284,102 @@ def test_should_return_1_if_never_expected_to_be_positive(self) -> None:
...
+                },
+            ),
+        ],
+        ids=["untagged_table"],
+    )
+    def test_should_raise_if_table_is_not_tagged(self, table: Table) -> None:
+        with pytest.raises(UntaggedTableError):
+            DummyClassifier().recall(table)  # type: ignore[arg-type]
+
+class TestF1_score:
```
Comment (lars-reimann) [inline tests/safeds/ml/classical/classification/test_classifier.py:337]:
```suggestion
class TestF1Score:
```

File: src/safeds/ml/classical/classification/_classifier.py [hunk 3]
```diff
@@ -136,3 +136,80 @@ def precision(self, validation_or_test_set: TaggedTable, positive_class: int = 1
...
+        Parameters
+        ----------
+        validation_or_test_set : TaggedTable
+            The validation or test set.
+        positive_class : int | str
+            The class to be considered positive. All other classes are considered negative.
+
+        Returns
+        -------
+        $F_1$-score : float
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:190]:
```suggestion
        f1_score : float
```

File: src/safeds/ml/classical/classification/_classifier.py [hunk 4]
```diff
@@ -136,3 +136,80 @@ def precision(self, validation_or_test_set: TaggedTable, positive_class: int = 1
         if (n_true_positives + n_false_positives) == 0:
             return 1.0
         return n_true_positives / (n_true_positives + n_false_positives)
+
+    def recall(self, validation_or_test_set: TaggedTable, positive_class: int = 1) -> float:
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:140]:
```suggestion
    def recall(self, validation_or_test_set: TaggedTable, positive_class: Any = 1) -> float:
```

Need to import `typing.Any`.

Might also need to be changed for `precision`.

File: src/safeds/ml/classical/classification/_classifier.py [hunk 5]
```diff
@@ -136,3 +136,80 @@ def precision(self, validation_or_test_set: TaggedTable, positive_class: int = 1
...
+        """
+        Compute the classifier's recall on the given data.
+
+        Parameters
+        ----------
+        validation_or_test_set : TaggedTable
+            The validation or test set.
+        positive_class : int | str
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:148]:
```suggestion
        positive_class : Any
```

**Review by lars-reimann — CHANGES_REQUESTED** [review]

File: src/safeds/ml/classical/classification/_classifier.py [hunk 6]
```diff
@@ -136,3 +136,80 @@ def precision(self, validation_or_test_set: TaggedTable, positive_class: int = 1
...
+                if expected_value == positive_class:
+                    n_true_positives += 1
+            elif expected_value == positive_class:
+                n_false_negatives += 1
+
+        if (n_true_positives + n_false_negatives) == 0:
+            return 1.0
+        return n_true_positives / (n_true_positives + n_false_negatives)
+
+    def f1_score(self, validation_or_test_set: TaggedTable, positive_class: int = 1) -> float:
```
Comment (lars-reimann) [inline src/safeds/ml/classical/classification/_classifier.py:177]:
```suggestion
    def f1_score(self, validation_or_test_set: TaggedTable, positive_class: Any = 1) -> float:
```

**Review by lars-reimann — APPROVED** [review]
